### PR TITLE
provide better error message when the content type is unexpected

### PR DESCRIPTION
**Pull request by rogpeppe**
This could be considered overkill but it will potentially help quite a bit.

### CODE REVIEW

File: checkisjson.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+		}
+	}
+}
+
+// saneText tries to make the given string easier to read when presented
+// as a single line. It squashes each run of white space into a single
+// space, trims leading and trailing white space and trailing full
+// stops. If newlineSemi is true, any newlines will be replaced with a
+// semicolon.
+func saneText(s string, newlineSemi bool) []byte {
```
Comment (mhilton) [inline checkisjson.go]: maybe call it sanitizeText (shrug)?

File: checkisjson.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	out := make([]byte, 0, len(s))
+	prevWhite := false
+	for _, r := range s {
```
Comment (mhilton) [inline checkisjson.go:131]: What does this do if the content isn't utf-8? I answered my own question http://play.golang.org/p/Hm4qAlwUvu it uses the replacement character, so its all good.

File: checkisjson.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+			// can fail is if there's a read error and we've
+			// removed that possibility by using
+			// noErrorReader above.
+			return errgo.Notef(err, "unexpected (and invalid) content text/html; want application/json; content: %q", sizeLimit(bodyData))
+		}
+		if len(text) == 0 {
+			return errgo.Newf(`unexpected content type text/html; want application/json; content: %q`, sizeLimit(bodyData))
+		}
+		return errgo.Newf(`unexpected content type text/html; want application/json; content: %s`, sizeLimit(text))
+	case "text/plain":
```
Comment (mhilton) [inline checkisjson.go:59]: is there a case for doing this with "text/*"?

File: checkisjson_test.go [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+</script>
+
+
+<!-- {version: ['0.1.29', '']} -->
+
+</body>
+</html>
+`,
+	expectError: `unexpected content type text/html; want application/json; content: Page not found | Juju; Jump to content; Store; Demo; About; Features; Community; Docs; Get started; ☰; Create; \+; 404: Sorry, we couldn’t find the page; Try a different URL, try searching for solutions or learn how to; create your own solution; Browse the store; All bundles; All charms; Submit a bug; Browse the store ›; Back to the top; Demo; About; Features; Docs; Get Started; Juju on Google+; Ubuntu Cloud on Twitter; Ubuntu Cloud on Facebook; © 2015 Canonical Ltd. Ubuntu and Canonical are registered trademarks of Canonical Ltd; Legal information; Report a bug on this site; Got to the top of the page`,
+}}
```
Comment (mhilton) [inline checkisjson_test.go:435]: Maybe add a test with the body in an 8-bit character set?

File: export_test.go [hunk 5]
```diff
@@ -1,3 +1,5 @@
 package httprequest
 
 var AppendURL = appendURL
+var CheckIsJSON = checkIsJSON
+var MaxErrorBodySize = &maxErrorBodySize
```
Comment (mhilton) [inline export_test.go]: newline at end of file.

File: unmarshal.go [hunk 6]
```diff
@@ -141,7 +139,7 @@ func unmarshalString(tag tag) unmarshaler {
 // unmarshalBody unmarshals the http request body
 // into the given value.
 func unmarshalBody(v reflect.Value, p Params, makeResult resultMaker) error {
-	if err := checkIsJSON(p.Request.Header); err != nil {
+	if err := checkIsJSON(p.Request.Header, p.Request.Body); err != nil {
```
Comment (mhilton) [inline unmarshal.go:142]: checkIsJSON(p.Request) [shrug]?

**Comment by mhilton** [comment]
LGTM with a few thoughts and nit-picks.

File: unmarshal.go (hunk 6 above)
Comment (rogpeppe) [inline unmarshal.go:142]: can't - we call it on responses too.

File: checkisjson.go (hunk 3 above)
Comment (rogpeppe) [inline checkisjson.go:59]: I think it's reasonable to quote non-plain text, such as xml.

File: export_test.go (hunk 5 above)
Comment (rogpeppe) [inline export_test.go]: done

File: checkisjson.go (hunk 1 above)
Comment (rogpeppe) [inline checkisjson.go]: done

File: checkisjson_test.go (hunk 4 above)
Comment (rogpeppe) [inline checkisjson_test.go:435]: done

File: checkisjson.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+	default:
+		return errgo.Newf(`unexpected content type %s; want application/json; content: %q`, mediaType, sizeLimit(bodyData))
+	}
+}
+
+type noErrorReader struct {
+	r io.Reader
+}
+
+func (r noErrorReader) Read(buf []byte) (int, error) {
```
Comment (frankban) [inline checkisjson.go:72]: I think this implementation deserves a docstring.

File: checkisjson_test.go [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+	about       string
+	contentType string
+	body        string
+	expectError string
+}{{
+	about:       "simple json",
+	contentType: "application/json; charset=UTF-8",
+}, {
+	about:       "simple json with charset",
+	contentType: "application/json; charset=UTF-8",
```
Comment (frankban) [inline checkisjson_test.go:27]: The first test has charset too, so those two are equal.

File: checkisjson_test.go [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+}, {
+	about:       "plain text with leading newline",
+	contentType: "text/plain; charset=UTF-8",
+	body:        "\nsome text",
+	expectError: `unexpected content type text/plain; want application/json; content: some text`,
+}, {
+	about:       "unknown content type",
+	contentType: "something",
+	body:        `some stuff`,
+	expectError: `unexpected content type something; want application/json; content: "some stuff"`,
```
Comment (frankban) [inline checkisjson_test.go]: I'd demonstrate that some stuff newlines or spaces are not modified in this case.

File: checkisjson_test.go [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+}, {
+	about:       "bad content type",
+	contentType: "/; charset=foo",
+	body:        `some stuff`,
+	expectError: `unexpected content type "/; charset=foo"; want application/json; content: "some stuff"`,
+}, {
+	about:       "large text body",
+	contentType: "text/plain",
+	body:        strings.Repeat("x", 1324),
```
Comment (frankban) [inline checkisjson_test.go]: 1024 + 300? [shrug]

File: checkisjson_test.go [hunk 11]
```diff
@@ -0,0 +1,453 @@
...
+</head><body>
+<h1>Proxy Error</h1>
+<p>The proxy server received an invalid
+response from an upstream server.<br />
+The proxy server could not handle the request <em><a href="/identity/v1/wait">GET&nbsp;/identity/v1/wait</a></em>.<p>
+Reason: <strong>Error reading from remote server</strong></p></p>
+<hr>
+<address>Apache/2.4.7 (Ubuntu) Server at api.jujucharms.com Port 443</address>
+</body></html>`,
+	expectError: `unexpected content type text/html; want application/json; content: 502 Proxy Error; Proxy Error; The proxy server received an invalid response from an upstream server; The proxy server could not handle the request; GET /identity/v1/wait; Reason:; Error reading from remote server; Apache/2\.4\.7 \(Ubuntu\) Server at api.jujucharms.com Port 443`,
```
Comment (frankban) [inline checkisjson_test.go:81]: This is awesome!

**Comment by frankban** [comment]
Great improvement Roger, thank you!
:+1: with only minor requests.

File: checkisjson.go (hunk 7 above)
Comment (rogpeppe) [inline checkisjson.go:72]: done

File: checkisjson_test.go (hunk 8 above)
Comment (rogpeppe) [inline checkisjson_test.go:27]: done

File: checkisjson_test.go (hunk 9 above)
Comment (rogpeppe) [inline checkisjson_test.go]: done

File: checkisjson_test.go (hunk 10 above)
Comment (rogpeppe) [inline checkisjson_test.go]: done

**Pull request merged**